### PR TITLE
[IR] Add an IR Builder with some basic functions

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue = #2193 

This PR renames the existing `IRBuilder` class to `ASTBuilder`, since it actually builds the frontend AST in C++.

The `IRBuilder` newly added in this PR is able to build the following instructions now:
- Kernel argument loads (I'm not sure if exposing the `DataType dt, bool is_ptr` arguments is a good idea)
- (Some) Constants
- (Some) Binary operations
- Prints

However, we cannot use the new `IRBuilder` to build a whole program now, since the constructor `Kernel::Kernel()` fetches the IR from `ASTBuilder`. I'm not sure how to solve it -- would it be fine to simply add a constructor to the `Kernel` class, and let developers call it to use the new `IRBuilder`?

And... the new `IRBuilder` is not unit-tested now.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/ir/ir_builder.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  Stmt *insert(std::unique_ptr<Stmt> &&stmt);
+
+  // Constants. TODO: add more types
+  Stmt *get_int32(int32 value);
+  Stmt *get_int64(int64 value);
+  Stmt *get_float32(float32 value);
+  Stmt *get_float64(float64 value);
+
+  // Kernel arguments.
+  Stmt *get_argument(int arg_id, DataType dt, bool is_ptr);
```
Comment (k-ye) [inline taichi/ir/ir_builder.h:31]: nit: `create_arg_load()`?

File: taichi/ir/ir_builder.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include "taichi/ir/ir.h"
+
+TLANG_NAMESPACE_BEGIN
+
+class IRBuilder {
+ public:
+  struct InsertPoint {
```
Comment (k-ye) [inline taichi/ir/ir_builder.h:9]: nit: does this need to be public?

File: taichi/ir/ir_builder.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  // l / r in C++
+  Stmt *create_div(Stmt *l, Stmt *r);
+  // floor(1.0 * l / r) in C++
+  Stmt *create_floordiv(Stmt *l, Stmt *r);
+  // 1.0 * l / r in C++
+  Stmt *create_truediv(Stmt *l, Stmt *r);
+
+  // Print values and strings. Arguments can be Stmt* or std::string.
+  template <typename... Args>
+  Stmt *create_print(Args &&... args);
```
Comment (k-ye) [inline taichi/ir/ir_builder.h:47]: I don't think you can put template impl inside CPP file?

**Review by k-ye — COMMENTED** [review]
Thanks! A few nits

File: taichi/ir/ir_builder.h (hunk 2 above)
Comment (xumingkuan) [reply taichi/ir/ir_builder.h:9]: I'm not sure if we need to modify the insertion point... I think it's fine to make it private now.

File: taichi/ir/ir_builder.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  Stmt *create_div(Stmt *l, Stmt *r);
+  // floor(1.0 * l / r) in C++
+  Stmt *create_floordiv(Stmt *l, Stmt *r);
+  // 1.0 * l / r in C++
+  Stmt *create_truediv(Stmt *l, Stmt *r);
+
+  // Print values and strings. Arguments can be Stmt* or std::string.
+  template <typename... Args>
+  Stmt *create_print(Args &&... args) {
+    return insert(Stmt::make<PrintStmt>(std::forward(args)));
```
Comment (k-ye) [inline taichi/ir/ir_builder.h:47]: Should this be `std::forward(args)...`?

**Review by k-ye — APPROVED** [review]
Looks like a great first step!

File: taichi/ir/ir_builder.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  InsertPoint insert_point_;
+
+ public:
+  IRBuilder();
+
+  // General inserter. Returns stmt.get().
+  Stmt *insert(std::unique_ptr<Stmt> &&stmt);
+
+  // Constants. TODO: add more types
+  Stmt *get_int32(int32 value);
```
Comment (yuanming-hu) [inline taichi/ir/ir_builder.h:24]: You may also consider a `template <typename T> Stmt *get_const(T val)` template function and then specialize for different types (or use `if constexpr (std::is_same_v<T, int32>)`). Just a random idea and the decision is yours :-)

**Review by yuanming-hu — APPROVED** [review]
Great! LGTM.

> However, we cannot use the new `IRBuilder` to build a whole program now, since the constructor `Kernel::Kernel()` fetches the IR from `ASTBuilder`. I'm not sure how to solve it -- would it be fine to simply add a constructor to the `Kernel` class, and let developers call it to use the new `IRBuilder`?

Your solution sounds good!

**Pull request merged**